tests/unit/test_utils.py

```diff
@@ -3476,3 +3477,29 @@ def cached_fn(self, a, b):
     assert cls2.cached_fn.cache_info().currsize == 2
     assert cls2.cached_fn.cache_info().hits == 1  # the call was a cache hit
     assert cls2.cached_fn.cache_info().misses == 2
+
+
+@pytest.fixture
+def operation_def():
+    return {'name': 'CreateFoo'}
+
+
+@pytest.mark.parametrize(
+    'request_dict, protocol, signature_version, expected_body',
+    [
+        ({'body': {'foo': 'bar'}}, 'query', 'v4', b'foo=bar'),
+        ({'body': b''}, 'query', 'v4', b''),
+        ({'body': {'foo': 'bar'}}, 'json', 'v4', {'foo': 'bar'}),
+        ({'body': {'foo': 'bar'}}, 'query', 'v2', {'foo': 'bar'}),
+        ({'body': 'foo=bar'}, 'query', 'v4', 'foo=bar'),
```

Comment:
At the risk of revealing my ignorance here: Why is the expected body here a `str` whereas in line 3490 is is `bytes`?

Reply:
They should both be strings I believe. Related issue in [this comment](https://github.com/boto/botocore/pull/2959#discussion_r1264245663) would make the tests consistent.